pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,296 @@
+package distributor
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/weaveworks/common/mtime"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	longPollDuration = 10 * time.Second
+)
+
+// ErrDontOverwrite is used when we want to accept a sample but not change the values in the KVStore when CAS'ing.
+type ErrDontOverwrite struct {
+}
+
+func (e *ErrDontOverwrite) Error() string {
+	return fmt.Sprintf("don't overwrite")
+}
+
+// Track the most recent timestamp for the replica we're
+// accepting samples from in a Prometheus HA setup.
+type replicaTracker struct {
+	replica   string
+	timestamp time.Time
+}
+
+// Track the replica we're accepting samples from
+// for each HA cluster we know about.
+type tracker struct {
+	logger log.Logger
+	cfg    HATrackerConfig
+
+	// Replicas we are accepting samples from.
+	electedLock sync.RWMutex
+	elected     map[string]replicaTracker
+	done        chan struct{}
+	cancel      context.CancelFunc
+
+	// We should only update the timestamp this often.
+	writeTimeout time.Duration
+	// Timeout that has to occur before we overwrite to a different replica.
+	overwriteTimeout time.Duration
+	client           ring.KVClient
+}
+
+// HATrackerConfig contains the configuration require to
+// create a HA Tracker.
+type HATrackerConfig struct {
+	ReplicaLabel string
+	ClusterLabel string
+
+	KVStore ring.KVConfig
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *HATrackerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.ReplicaLabel, "ha-tracker.replica", "replica", "Prometheus label to look for in samples to identify a Proemtheus HA replica.")
```

Comment:
Since this label is for internal use by cortex can we default to something like `__replica`? That way it doesn't interfere with any series that already use this label.

Reply:
I could be not interpreting this correctly, but this is the external label Prometheus is writing correct? Similar to `--query.replica-label` in Thanos? If so, has there been thought about making this configurable per customer for people with different configurations?

As far as defaults go, I vote `prometheus_replica` since that is what the Prometheus operator uses by default.

Reply:
Correct, this is the external label you would add to your Prometheus config. You're right it probably should be configurable per customer. At the same time, I'm thinking we can actually just drop both the cluster and replica label before storing the data.

Reply:
I think replica label we should definitely drop, otherwise you would have some weird disjointed series whenever you switch replicas.

I don't think cluster should be dropped since it is common for multiple clusters to send to a cortex and would need to be distinguished. Speaking of cluster, if someone only has one cluster but wants to use deduping they would have to specify the cluster label correct?

Reply:
:+1:, looks like Thanos defaults to `replica` for this label.  I do think we should use `__replica__` or `__prometheus_replica__`, as the change a random timeseries has a `replica` label is quite high.

Reply:
Lets leave as `replica`, and make this configurable per user.